File: csharp/Princeton-Algorithms1/Week01/Percolation/PercolationStats.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Princeton_Algorithms1.Common;

namespace Princeton_Algorithms1.Week01.Percolation
{
    public sealed class PercolationStats
    {
        private double[] results;
        private readonly int _totalNumberOfElements;

        public Func<IUnionFindImplementation> UnionFindImplementationFactory { get; set; }
            = () => new WeightedPathCompressingQuickUnionFind();

        /// <summary>
        ///     Performs independent trials on an n-by-n grid
        /// </summary>
        /// <param name="n"></param>
        /// <param name="trials"></param>
        public PercolationStats(int n, int trials)
        {
            if (n <= 0)
                throw new ArgumentOutOfRangeException(nameof(n), n, "Value cannot be less than 1");
            if (trials <= 0)
                throw new ArgumentOutOfRangeException(nameof(trials), trials, "Value cannot be less than 1");

            _totalNumberOfElements = n * n;
            RunTrials(n, trials);
        }

        private void RunTrials(int n, int trials)
        {
            results = new double[trials];
            Parallel.For(
                fromInclusive: 0,
                toExclusive: trials,
                body: index =>
                {
                    results[index] = RunTrial(n);
                }
            );
        }

        private double RunTrial(int n)
        {
            Percolation percolator = new Percolation(n, UnionFindImplementationFactory());
            while (!percolator.Percolates())
            {
                int row = ThreadLocalRandom.Next(n); // TODO: move random to dependency
                int col = ThreadLocalRandom.Next(n); // TODO: move random to dependency
                percolator.Open(row, col);
            }
            double openCount = percolator.NumberOfOpenSites();
            return openCount / (_totalNumberOfElements);
        }

        // sample mean of percolation threshold
        public double Mean()
        {
            return results.Average();
        }

        // sample standard deviation of percolation threshold
        public double Stddev()
        {
            throw new NotImplementedException();
        }

        // low endpoint of 95% confidence interval
        public double ConfidenceLo()
        {
            throw new NotImplementedException();
        }

        // high endpoint of 95% confidence interval
        public double ConfidenceHi()
        {
            throw new NotImplementedException();
        }
    }
}

File: java/percolation.princeton-algorithms1.arphox/src/PercolationStats.java
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.StdStats;

public class PercolationStats {
    private final int trials;
    private final double[] results;
    private final double meanValue;
    private final double confidence;

    // perform independent trials on an n-by-n grid
    public PercolationStats(int n, int trials) {
        if (n < 1) throw new IllegalArgumentException("Value of n is invalid.");
        if (trials < 1) throw new IllegalArgumentException("Value of trials is invalid.");

        results = new double[trials];
        this.trials = trials;
        int totalNumberOfElements = n * n;
        for (int i = 0; i < trials; i++) {
            double openCount = runTrial(n);
            results[i] = openCount / totalNumberOfElements;
        }
        meanValue = StdStats.mean(results);
        confidence = calculateConfidence();
    }

    // test client (see below)
    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        int trials = Integer.parseInt(args[1]);

        PercolationStats stats = new PercolationStats(n, trials);
        String confidence = String.format("[%s, %s]", stats.confidenceLo(), stats.confidenceHi());
        StdOut.println("mean                    = " + stats.mean());
        StdOut.println("stddev                  = " + stats.stddev());
        StdOut.println("95% confidence interval = " + confidence);
    }

    private double runTrial(int n) {
        Percolation percolator = new Percolation(n);
        while (!percolator.percolates()) {
            int row = StdRandom.uniform(n) + 1;
            int col = StdRandom.uniform(n) + 1;
            percolator.open(row, col);
        }
        return percolator.numberOfOpenSites();
    }

    // sample mean of percolation threshold
    public double mean() {
        return meanValue;
    }

    // sample standard deviation of percolation threshold
    public double stddev() {
        return StdStats.stddev(results);
    }

    // low endpoint of 95% confidence interval
    public double confidenceLo() {
        return mean() - confidence;
    }

    // high endpoint of 95% confidence interval
    public double confidenceHi() {
        return mean() + confidence;
    }

    private double calculateConfidence() {
        double sum = 0;
        double mean = mean();
        for (double result : results) {
            double notSquared = result - mean;
            double squared = notSquared * notSquared;
            sum += squared;
        }

        double sSquared = sum / (trials - 1);
        double s = Math.sqrt(sSquared);
        double sqrtTrials = Math.sqrt(trials);
        return (1.96 * s) / sqrtTrials;
    }
}